### PR TITLE
Pipeline jenkin job for upstream build

### PR DESCRIPTION
**Pull request by anrao19**
Signed-off-by: Anuchaithra Rao <anrao@redhat.com>

# Description
https://ceph-downstream-jenkins-csb-storage.apps.ocp-c1.prod.psi.redhat.com/job/cephci-upstream-test-executor/12/console

latest console: 
https://ceph-downstream-jenkins-csb-storage.apps.ocp-c1.prod.psi.redhat.com/job/cephci-upstream-test-executor/7/console
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-D9OY12

Updated pr Console: https://ceph-downstream-jenkins-csb-storage.apps.ocp-c1.prod.psi.redhat.com/job/upstream_suite_executor/11/console
EX: Logs: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-VQBMKW/

https://ceph-downstream-jenkins-csb-storage.apps.ocp-c1.prod.psi.redhat.com/job/upstream_test_executor/4/console
script (.sh) files added are example suites

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by psathyan** [comment]
Can we opt to using the metadata approach instead of shell scripts?

File: run.py [hunk 1]
```diff
@@ -444,10 +445,15 @@ def run(args):
...
-            build, rhbuild, platform
-        )
+    if build:
+        if build not in ["released", "upstream"]:
+            base_url, docker_registry, docker_image, docker_tag = fetch_build_artifacts(
+                build, rhbuild, platform
+            )
+        elif build == "upstream":
+            base_url, docker_registry, docker_image, docker_tag = fetch_upstream_build(
+                build, rhbuild
```
Comment (sunilangadi2) [inline run.py:455]: @anrao19 rhbuild [here](https://github.com/red-hat-storage/cephci/blob/fa710c14fb951f17e94f60bcc98e3eff1cb986a7/run.py#L112) mentioned it as ceph downstream version, Is it ok to use same vraible for upstream ceph version?

File: run.py [hunk 2]
```diff
@@ -444,9 +446,12 @@ def run(args):
     platform = args["--platform"]
     build = args.get("--build", None)
 
+    upstream_build = None
     if build and build not in ["released"]:
+        if build == "upstream":
+            upstream_build = args.get("--upstream-build", None)
```
Comment (sunilkumarn417) [inline run.py:452]: I think  `--upstream-build` would be mandatory if `build == "upstream"`.

File: utility/utils.py [hunk 3]
```diff
@@ -1109,33 +1109,37 @@ def generate_self_signed_certificate(subject: Dict) -> Tuple:
...
         build: build section to be fetched
         platform: OS distribution name with major Version(ex., rhel-8)
-
+        upstream_build: upstream build(ex., pacific/quincy) default:None
     Returns:
         base_url, container_registry, image-name, image-tag
     """
     recipe_url = get_cephci_config().get("build-url", magna_rhcs_artifacts)
-    url = f"{recipe_url}RHCEPH-{ceph_version}.yaml"
+    filename = f"{build}.yaml" if upstream_build else f"RHCEPH-{ceph_version}.yaml"
```
Comment (sunilkumarn417) [inline utility/utils.py:1126]: I think it is `build == "upstream"`, then `f"{build}.yaml`

**Review by sunilkumarn417 — COMMENTED** [review]
Thanks for getting this is.

Few minor comments.

File: pipeline/metadata/quincy.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+# Metadata file for quincy upstream testingn.
+# Single file to specify test suites to be executed for all regression (sanity) and schedule tests
+# as per defined in each stages.
+# Each stage will execute in sequential pattern.
+# Contains all Default parameter used for Execution when no override is specified.
+# parameter required to create a similar custom file is suite name, suite yaml file, global configuration file,
+# platform, rhbuild, inventory and metadata information like frequency of execution, tier, cloud type, functional group and stage.
+#=====================================================================================
+
+- name: "Beginning Red Hat Cephadm testing with upstream build"
```
Comment (sunilkumarn417) [inline pipeline/metadata/quincy.yaml:11]: name: "Beginning Red Hat Cephadm testing with upstream build" 

please reword to generic name.

File: run.py (hunk 2 above)
Comment (sunilkumarn417) [reply run.py:452]:
```
base_url, docker_registry, docker_image, docker_tag = fetch_build_artifacts(
            build, rhbuild, platform, args.get("--upstream-build", None)
        )
```

File: pipeline/metadata/quincy.yaml (hunk 4 above)
Comment (anrao19) [reply pipeline/metadata/quincy.yaml]: Made changes as per comments

**Review by sunilangadi2 — DISMISSED** [review]
Thanks, @anrao19 LGTM.

**Review by sunilkumarn417 — COMMENTED** [review]
LGTM

**Review by sunilkumarn417 — DISMISSED** [review]

File: utility/utils.py [hunk 5]
```diff
@@ -1109,35 +1109,47 @@ def generate_self_signed_certificate(subject: Dict) -> Tuple:
     )
 
 
-def fetch_build_artifacts(build, ceph_version, platform):
+def fetch_build_artifacts(build, ceph_version, platform, upstream_build):
```
Comment (psathyan) [inline utility/utils.py:1112]: I believe there is only one place which calls this method... but still feel we should set `upstream_build=None`.

File: suites/quincy/rgw/tier-0_rgw.yaml [hunk 6]
```diff
@@ -0,0 +1,123 @@
+# Tier 0: RGW build evaluation
```
Comment (psathyan) [inline suites/quincy/rgw/tier-0_rgw.yaml:1]: Could we rename the test suite filename that is more appropriate? The below tests look more like functional testing i.e. `Tier-1`

File: suites/quincy/cephadm/tier-0_dmfg.yaml [hunk 7]
```diff
@@ -0,0 +1,119 @@
+#======================================6=========================================================
```
Comment (psathyan) [inline suites/quincy/cephadm/tier-0_dmfg.yaml:1]: Do we need run these tests as part of build sanity check?

File: pipeline/upstream_test_executor.groovy [hunk 8]
```diff
@@ -0,0 +1,76 @@
+/*
+    Script that trigger testsuite for upstream build
+*/
+// Global variables section
+def nodeName = "centos-7"
```
Comment (psathyan) [inline pipeline/upstream_test_executor.groovy:5]: Could we move away from `centos-7`?

**Review by psathyan — COMMENTED** [review]
I think this is a good start but would like to some changes that would be forward looking. For example, 

- change the way we execute test suites
- we could be setting the path how we want the team to follow

File: ceph/ceph_admin/__init__.py [hunk 9]
```diff
@@ -155,6 +155,11 @@ def setup_upstream_repository(self, repo_url=None):
                 sudo=True, cmd=f"curl -o /etc/yum.repos.d/upstream_ceph.repo {repo_url}"
             )
             node.exec_command(sudo=True, cmd="yum update metadata", check_ec=False)
+            node.exec_command(
+                sudo=True,
+                cmd="dnf install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm -y",
+                check_ec=False,
+            )
```
[the remark this answers is not in the document]
Comment (anrao19) [reply ceph/ceph_admin/__init__.py:162]: We will take this up when there is a new OS veresion

File: suites/quincy/cephadm/tier-0_dmfg.yaml (hunk 7 above)
Comment (anrao19) [reply suites/quincy/cephadm/tier-0_dmfg.yaml]: As of now we will be running this along with other suites, if there is any changes required we will take this as an improvement

File: pipeline/upstream_test_executor.groovy (hunk 8 above)
Comment (anrao19) [reply pipeline/upstream_test_executor.groovy]: Made all changes as per comment

File: ceph/ceph_admin/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply ceph/ceph_admin/__init__.py:162]: We had discussion over this before implementation, and it would be in the backlog as an improvement.

**Review by sunilkumarn417 — COMMENTED** [review]
Review changes LGTM.

File: suites/quincy/cephfs/cephfs-basic-mds-operations.yaml [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      polarion-id: "CEPH-83574284"
+  - test:
+      abort-on-fail: false
+      desc: "cephfs basic operations"
+      module: cephfs_basic_tests.py
+      name: cephfs-basics
+      polarion-id: CEPH-11293  # also applies to [CEPH-11296,CEPH-11297,CEPH-11295]
+  - test:
+      name: Increase and Decrease of MDS
+      module: mds_Inc_dec.py
```
Comment (psathyan) [inline suites/quincy/cephfs/cephfs-basic-mds-operations.yaml:129]: Could we normalize the module name?

File: suites/quincy/cephfs/cephfs-basic-mds-operations.yaml [hunk 11]
```diff
@@ -0,0 +1,132 @@
+# Tier0: cephfs build evaluation
```
Comment (psathyan) [inline suites/quincy/cephfs/cephfs-basic-mds-operations.yaml:1]: nitpick... sorry
We are already in `cephfs` folder... could we avoid the prefix to the filename? IMO, the filename could be `test-core-functionality.yaml`

File: suites/quincy/rbd/rbd-upstream-cli-operations.yaml [hunk 12]
```diff
@@ -0,0 +1,150 @@
+# Tier0: RBD build evaluation
```
Comment (psathyan) [inline suites/quincy/rbd/rbd-upstream-cli-operations.yaml:1]: Likewise, IMO the filename could be `basic-operations_unit_testing.yaml`

File: suites/quincy/rgw/rgw-basic-bucket-object-operations.yaml [hunk 13]
```diff
@@ -0,0 +1,123 @@
+# RGW build evaluation
```
Comment (psathyan) [inline suites/quincy/rgw/rgw-basic-bucket-object-operations.yaml:1]: likewise, IMO the filename could be `test-basic-object-operations.yaml`

**Review by psathyan — COMMENTED** [review]
Thanks... overall, I believe we are ready to get this merged

File: suites/quincy/cephfs/cephfs-basic-mds-operations.yaml (hunk 10 above)
Comment (anrao19) [reply suites/quincy/cephfs/cephfs-basic-mds-operations.yaml]: Created an Issue: https://github.com/red-hat-storage/cephci/issues/1558

File: suites/quincy/rbd/rbd-upstream-cli-operations.yaml (hunk 12 above)
Comment (anrao19) [reply suites/quincy/rbd/rbd-upstream-cli-operations.yaml]: Updated as per comments

**Review by psathyan — COMMENTED** [review]
Looks good to me

**Review by sunilkumarn417 — APPROVED** [review]
Thanks @anrao19 for completing this workflow.
Thanks @psathyan.

LGTM++

**Pull request merged**